experimental/components/src/units/converter_factory.rs

```diff
@@ -0,0 +1,339 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+use litemap::LiteMap;
+use num_bigint::BigInt;
+use num_rational::Ratio;
+use num_traits::{One, Zero};
+use zerotrie::ZeroTrieSimpleAscii;
+use zerovec::ule::AsULE;
+use zerovec::{ZeroSlice, ZeroVec};
+
+use crate::units::provider::{MeasureUnitItem, SignULE};
+
+use super::{
+    converter::{
+        OffsetConverter, ProportionalConverter, ReciprocalConverter, UnitsConverter,
+        UnitsConverterInner,
+    },
+    measureunit::{MeasureUnit, MeasureUnitParser},
+    provider::{Base, SiPrefix, Sign, UnitsInfoV1},
+};
+
+/// ConverterFactory is a factory for creating a converter.
+pub struct ConverterFactory<'data> {
+    // TODO(#4522): Make the converter factory owns the data.
+    /// Contains the necessary data for the conversion factory.
+    payload: &'data UnitsInfoV1<'data>,
+    payload_store: &'data ZeroTrieSimpleAscii<ZeroVec<'data, u8>>,
+}
+
+impl From<Sign> for num_bigint::Sign {
+    fn from(val: Sign) -> Self {
+        match val {
+            Sign::Positive => num_bigint::Sign::Plus,
+            Sign::Negative => num_bigint::Sign::Minus,
+        }
+    }
+}
+
+impl<'data> ConverterFactory<'data> {
+    #[cfg(feature = "datagen")]
+    pub fn from_payload(
+        payload: &'data UnitsInfoV1<'data>,
+        payload_store: &'data ZeroTrieSimpleAscii<ZeroVec<'data, u8>>,
+    ) -> Self {
+        Self {
+            payload,
+            payload_store,
+        }
+    }
+
+    pub fn parser(&self) -> MeasureUnitParser<'_> {
+        MeasureUnitParser::from_payload(self.payload_store)
+    }
+
+    /// Computes the offset between the input and output units.
+    /// To calculate the offset, we follow these steps:
+    /// 1. Assume the conversion rate for unit 1 to the root is: N1/D1 + OffsetN1/OffsetD1.
+    /// 2. Assume the conversion rate for unit 2 to the root is: N2/D2 + OffsetN2/OffsetD2.
+    /// 3. To convert from the root to unit 2, the formula is: D2/N2 - OffsetN2/OffsetD2 * (D2/N2).
+    /// 4. Given a value V to be converted from unit 1 to unit 2, the conversion will be the root:
+    ///    (V * N1/D1) + OffsetN1/OffsetD1, denoted as V_Root.
+    /// 5. To convert V_Root to unit 2, the formula is: V_Root * D2/N2 - OffsetN2/OffsetD2 * D2/N2.
+    /// 6. Substituting V_Root from step 4 into step 5 yields:
+    ///    ((V * N1/D1) + OffsetN1/OffsetD1) * D2/N2 - OffsetN2/OffsetD2 * (D2/N2).
+    /// 7. Simplifying the equation gives:
+    ///    V * (N1/D1) * (D2/N2) + OffsetN1/OffsetD1 * (D2/N2) - OffsetN2/OffsetD2 * (D2/N2).
+    /// 8. Focusing on the constants (offsets), we derive the offset formula:
```

Comment:
I think there are 5 steps too many here. You have two assumptions, and this formula should follow from them with less arithmetic than you're using here

Reply:
I have simplify the illustration.